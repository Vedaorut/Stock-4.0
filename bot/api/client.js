import axios from 'axios';
import dotenv from 'dotenv';

dotenv.config();

/**
 * Base HTTP client with retry logic and error handling
 */
class ApiClient {
  constructor() {
    this.baseURL = process.env.BACKEND_URL || 'http://localhost:3000';
    this.timeout = parseInt(process.env.API_TIMEOUT) || 10000;

    this.client = axios.create({
      baseURL: this.baseURL,
      timeout: this.timeout,
      headers: {
        'Content-Type': 'application/json'
      }
    });

    this._setupInterceptors();
  }

  /**
   * Setup request/response interceptors
   */
  _setupInterceptors() {
    // Response interceptor with retry logic
    this.client.interceptors.response.use(
      response => response,
      async (error) => {
        const config = error.config;

        // Retry logic for server errors (500+)
        if (error.response?.status >= 500 && (!config.retryCount || config.retryCount < 3)) {
          config.retryCount = (config.retryCount || 0) + 1;

          console.warn(
            `üîÑ Retry ${config.retryCount}/3 for ${config.method.toUpperCase()} ${config.url}`
          );

          // Exponential backoff: 1s, 2s, 4s
          const delay = Math.pow(2, config.retryCount - 1) * 1000;
          await new Promise(resolve => setTimeout(resolve, delay));

          return this.client.request(config);
        }

        throw error;
      }
    );
  }

  /**
   * Make HTTP request
   * @param {string} method - HTTP method
   * @param {string} url - Endpoint URL
   * @param {object} data - Request body
   * @param {string} token - JWT token
   * @returns {Promise<object>} Response data
   */
  async request(method, url, data = null, token = null) {
    try {
      const config = {
        method,
        url,
        data
      };

      // Add authorization header if token provided
      if (token) {
        config.headers = {
          'Authorization': `Bearer ${token}`
        };
      }

      const response = await this.client.request(config);
      return response.data;
    } catch (error) {
      console.error(
        `‚ùå API ${method.toUpperCase()} ${url} failed:`,
        error.response?.data || error.message
      );
      throw this._handleError(error);
    }
  }

  /**
   * Handle and format errors
   * @param {Error} error - Axios error
   * @returns {Error} Formatted error
   */
  _handleError(error) {
    if (error.response) {
      // Server responded with error status
      const status = error.response.status;
      const message = error.response.data?.error || error.response.data?.message;

      if (status === 401) {
        return new Error('–û—à–∏–±–∫–∞ –∞–≤—Ç–æ—Ä–∏–∑–∞—Ü–∏–∏. –ü–æ–ø—Ä–æ–±—É–π—Ç–µ –≤–æ–π—Ç–∏ –∑–∞–Ω–æ–≤–æ.');
      }
      if (status === 403) {
        return new Error('–î–æ—Å—Ç—É–ø –∑–∞–ø—Ä–µ—â–µ–Ω.');
      }
      if (status === 404) {
        return new Error(message || '–†–µ—Å—É—Ä—Å –Ω–µ –Ω–∞–π–¥–µ–Ω.');
      }
      if (status === 429) {
        return new Error('–°–ª–∏—à–∫–æ–º –º–Ω–æ–≥–æ –∑–∞–ø—Ä–æ—Å–æ–≤. –ü–æ–ø—Ä–æ–±—É–π—Ç–µ –ø–æ–∑–∂–µ.');
      }
      if (status >= 500) {
        return new Error('–û—à–∏–±–∫–∞ —Å–µ—Ä–≤–µ—Ä–∞. –ü–æ–ø—Ä–æ–±—É–π—Ç–µ –ø–æ–∑–∂–µ.');
      }

      return new Error(message || '–û—à–∏–±–∫–∞ –∑–∞–ø—Ä–æ—Å–∞.');
    }

    if (error.request) {
      // Request was made but no response received
      return new Error('–û—à–∏–±–∫–∞ —Å–µ—Ç–∏. –ü—Ä–æ–≤–µ—Ä—å—Ç–µ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ –∏–Ω—Ç–µ—Ä–Ω–µ—Ç—É.');
    }

    // Something else happened
    return new Error(error.message || '–ù–µ–∏–∑–≤–µ—Å—Ç–Ω–∞—è –æ—à–∏–±–∫–∞.');
  }

  /**
   * GET request helper
   */
  async get(url, token = null) {
    return this.request('GET', url, null, token);
  }

  /**
   * POST request helper
   */
  async post(url, data, token = null) {
    return this.request('POST', url, data, token);
  }

  /**
   * PUT request helper
   */
  async put(url, data, token = null) {
    return this.request('PUT', url, data, token);
  }

  /**
   * DELETE request helper
   */
  async delete(url, token = null) {
    return this.request('DELETE', url, null, token);
  }
}

// Export singleton instance
export default new ApiClient();
