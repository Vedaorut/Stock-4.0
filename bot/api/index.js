/**
 * API Client for Status Stock Backend
 *
 * Centralized API integration for Telegram Bot
 * All endpoints use JWT authentication and include retry logic
 */

export { authApi } from './auth.js';
export { shopsApi } from './shops.js';
export { productsApi } from './products.js';
export { ordersApi } from './orders.js';
export { subscriptionsApi } from './subscriptions.js';
export { paymentsApi } from './payments.js';

// Export base client for custom requests
export { default as apiClient } from './client.js';
