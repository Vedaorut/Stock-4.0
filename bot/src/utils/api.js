import axios from 'axios';
import config from '../config/index.js';
import logger from './logger.js';

// Create axios instance with base URL
const api = axios.create({
  baseURL: config.backendUrl + '/api',
  timeout: 10000,
  headers: {
    'Content-Type': 'application/json'
  }
});

// Request interceptor for logging
api.interceptors.request.use(
  (config) => {
    logger.debug(`API Request: ${config.method.toUpperCase()} ${config.url}`);
    return config;
  },
  (error) => {
    logger.error('API Request Error:', error);
    return Promise.reject(error);
  }
);

// Response interceptor for error handling
api.interceptors.response.use(
  (response) => {
    logger.debug(`API Response: ${response.status} ${response.config.url}`);
    return response;
  },
  (error) => {
    if (error.response) {
      // Log full error details including validation errors and request body
      let requestBody = null;
      if (error.config?.data) {
        if (typeof error.config.data === 'string') {
          try {
            requestBody = JSON.parse(error.config.data);
          } catch {
            requestBody = error.config.data;
          }
        } else {
          requestBody = error.config.data;
        }
      }

      logger.error(`API Error: ${error.response.status} ${error.response.config.url}`, {
        responseData: error.response.data,
        requestBody,
        validationErrors: error.response.data?.details || null
      });
    } else if (error.request) {
      logger.error('API Error: No response received', { url: error.config?.url });
    } else {
      logger.error('API Error:', error.message);
    }
    return Promise.reject(error);
  }
);

// API methods
export const authApi = {
  // Register or login user
  async authenticate(telegramId, userData) {
    const requestBody = {
      telegramId: parseInt(telegramId, 10),  // Send as integer, not string
      username: userData.username,
      firstName: userData.firstName || userData.first_name,  // Support both camelCase and snake_case
      lastName: userData.lastName || userData.last_name || ''
    };

    const { data } = await api.post('/auth/register', requestBody, {
      headers: {
        'Content-Type': 'application/json'
      }
    });
    // Unwrap response: return data.data instead of data
    return data.data || data;
  },

  // Update user role
  async updateRole(role, token) {
    const { data } = await api.patch('/auth/role',
      { role },
      {
        headers: { Authorization: `Bearer ${token}` }
      }
    );
    return data.data || data;
  }
};

export const shopApi = {
  // Get user's shop
  async getMyShop(token) {
    const { data } = await api.get('/shops/my', {
      headers: { Authorization: `Bearer ${token}` }
    });
    const shops = data.data || data;
    return Array.isArray(shops) ? shops : [];
  },

  // Create new shop
  async createShop(shopData, token) {
    const { data } = await api.post('/shops', shopData, {
      headers: { Authorization: `Bearer ${token}` }
    });
    // Unwrap response: return data.data (shop object) instead of wrapper
    return data.data || data;
  },

  // Search shops
  async searchShops(query, token = null) {
    const config = {
      params: { q: query }
    };

    if (token) {
      config.headers = { Authorization: `Bearer ${token}` };
    }

    const { data } = await api.get('/shops/search', config);
    const shops = data.data || data;
    return Array.isArray(shops) ? shops : [];
  },

  // Get shop by ID
  async getShop(shopId) {
    const { data } = await api.get(`/shops/${shopId}`);
    // Unwrap response: return data.data (shop object) instead of wrapper
    return data.data || data;
  }
};

export const productApi = {
  // Create product
  async createProduct(productData, token) {
    const { data } = await api.post('/products', productData, {
      headers: { Authorization: `Bearer ${token}` }
    });
    // Unwrap response: return data.data (product object) instead of wrapper
    return data.data || data;
  },

  // Get shop products
  async getShopProducts(shopId) {
    const { data } = await api.get('/products', {
      params: { shopId }
    });
    // Unwrap response: return data.data (array of products) instead of wrapper
    return data.data || data;
  }
};

export const orderApi = {
  // Get buyer orders
  async getMyOrders(token) {
    const { data } = await api.get('/orders/my', {
      headers: { Authorization: `Bearer ${token}` }
    });
    // Unwrap response: return data.data (array of orders) instead of wrapper
    const orders = data.data || data;
    return Array.isArray(orders) ? orders : [];
  },

  // Get shop orders (sales)
  async getShopOrders(shopId, token) {
    const { data } = await api.get('/orders', {
      params: { shopId },
      headers: { Authorization: `Bearer ${token}` }
    });
    // Unwrap response: return data.data (array of orders) instead of wrapper
    const orders = data.data || data;
    return Array.isArray(orders) ? orders : [];
  }
};

export const paymentApi = {
  // Verify crypto payment
  async verifyPayment(paymentData, token) {
    const { data } = await api.post('/payments/verify', paymentData, {
      headers: { Authorization: `Bearer ${token}` }
    });
    // Unwrap response: return data.data (payment object) instead of wrapper
    return data.data || data;
  },

  // Generate crypto address
  async generateAddress(currency, token) {
    const { data } = await api.post('/payments/address',
      { currency },
      { headers: { Authorization: `Bearer ${token}` } }
    );
    // Unwrap response: return data.data (address object) instead of wrapper
    return data.data || data;
  }
};

export const subscriptionApi = {
  // Check if user is subscribed to shop
  async checkSubscription(shopId, token) {
    const { data } = await api.get(`/subscriptions/check/${shopId}`, {
      headers: { Authorization: `Bearer ${token}` }
    });
    // Unwrap response: return data.data (object with isSubscribed and subscription) instead of wrapper
    return data.data || data;
  },

  // Subscribe to shop
  async subscribe(shopId, token) {
    const { data } = await api.post(
      '/subscriptions',
      { shopId: Number(shopId) },
      {
        headers: { Authorization: `Bearer ${token}` }
      }
    );
    // Unwrap response: return data.data (subscription object) instead of wrapper
    return data.data || data;
  },

  // Get user subscriptions
  async getMySubscriptions(token) {
    const { data } = await api.get('/subscriptions', {
      headers: { Authorization: `Bearer ${token}` }
    });
    // Unwrap response: return data.data (array of subscriptions) instead of wrapper
    return data.data || data;
  },

  // Unsubscribe from shop
  async unsubscribe(shopId, token) {
    const { data } = await api.delete(`/subscriptions/${shopId}`, {
      headers: { Authorization: `Bearer ${token}` }
    });
    // Unwrap response: return data.data instead of wrapper
    return data.data || data;
  }
};

export const walletApi = {
  // Get shop wallets
  async getWallets(shopId, token) {
    const { data } = await api.get(`/shops/${shopId}/wallets`, {
      headers: { Authorization: `Bearer ${token}` }
    });
    return data.data || data;
  },

  // Update shop wallets
  async updateWallets(shopId, wallets, token) {
    const { data } = await api.put(
      `/shops/${shopId}/wallets`,
      wallets,
      {
        headers: { Authorization: `Bearer ${token}` }
      }
    );
    return data.data || data;
  }
};

// Export named api instance for testing
export { api };
export default api;
