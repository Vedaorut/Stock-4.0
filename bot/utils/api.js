import axios from 'axios';
import dotenv from 'dotenv';

dotenv.config();

const BACKEND_URL = process.env.BACKEND_URL || 'http://localhost:3000';
const API_TIMEOUT = parseInt(process.env.API_TIMEOUT) || 10000;

// Create axios instance with default config
const apiClient = axios.create({
  baseURL: BACKEND_URL,
  timeout: API_TIMEOUT,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Request interceptor for adding auth token
apiClient.interceptors.request.use(
  (config) => {
    // Token will be added per request if needed
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

// Response interceptor for error handling
apiClient.interceptors.response.use(
  (response) => response,
  (error) => {
    if (error.response) {
      // Server responded with error status
      console.error('API Error:', error.response.status, error.response.data);
    } else if (error.request) {
      // Request was made but no response received
      console.error('Network Error: No response from server');
    } else {
      // Something else happened
      console.error('Request Error:', error.message);
    }
    return Promise.reject(error);
  }
);

// User API functions
export async function createUser(telegramId, username, firstName, role) {
  try {
    const response = await apiClient.post('/api/users', {
      telegramId: telegramId.toString(),
      username,
      firstName,
      role, // 'buyer' or 'seller'
    });
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка создания пользователя' };
  }
}

export async function getUser(telegramId) {
  try {
    const response = await apiClient.get(`/api/users/telegram/${telegramId}`);
    return { success: true, data: response.data };
  } catch (error) {
    if (error.response?.status === 404) {
      return { success: false, notFound: true };
    }
    return { success: false, error: error.response?.data?.message || 'Ошибка получения данных пользователя' };
  }
}

export async function updateUser(telegramId, updates) {
  try {
    const response = await apiClient.patch(`/api/users/telegram/${telegramId}`, updates);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка обновления пользователя' };
  }
}

// Shop API functions
export async function createShop(telegramId, shopName, paymentHash) {
  try {
    const response = await apiClient.post('/api/shops', {
      telegramId: telegramId.toString(),
      name: shopName,
      paymentHash,
    });
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка создания магазина' };
  }
}

export async function getShopByOwner(telegramId) {
  try {
    const response = await apiClient.get(`/api/shops/owner/${telegramId}`);
    return { success: true, data: response.data };
  } catch (error) {
    if (error.response?.status === 404) {
      return { success: false, notFound: true };
    }
    return { success: false, error: error.response?.data?.message || 'Ошибка получения магазина' };
  }
}

export async function getShopByName(shopName) {
  try {
    const response = await apiClient.get(`/api/shops/search?name=${encodeURIComponent(shopName)}`);
    return { success: true, data: response.data };
  } catch (error) {
    if (error.response?.status === 404) {
      return { success: false, notFound: true };
    }
    return { success: false, error: error.response?.data?.message || 'Магазин не найден' };
  }
}

export async function updateShop(shopId, updates) {
  try {
    const response = await apiClient.patch(`/api/shops/${shopId}`, updates);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка обновления магазина' };
  }
}

// Product API functions
export async function createProduct(shopId, productData) {
  try {
    const response = await apiClient.post('/api/products', {
      shopId,
      ...productData,
    });
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка добавления товара' };
  }
}

export async function getProducts(shopId) {
  try {
    const response = await apiClient.get(`/api/products/shop/${shopId}`);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка получения товаров' };
  }
}

export async function updateProduct(productId, updates) {
  try {
    const response = await apiClient.patch(`/api/products/${productId}`, updates);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка обновления товара' };
  }
}

export async function deleteProduct(productId) {
  try {
    const response = await apiClient.delete(`/api/products/${productId}`);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка удаления товара' };
  }
}

// Order API functions
export async function getOrdersByShop(shopId) {
  try {
    const response = await apiClient.get(`/api/orders/shop/${shopId}`);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка получения заказов' };
  }
}

export async function getOrdersByBuyer(telegramId) {
  try {
    const response = await apiClient.get(`/api/orders/buyer/${telegramId}`);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка получения заказов' };
  }
}

export async function updateOrderStatus(orderId, status) {
  try {
    const response = await apiClient.patch(`/api/orders/${orderId}/status`, { status });
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка обновления статуса заказа' };
  }
}

// Subscription API functions
export async function subscribeToShop(telegramId, shopId) {
  try {
    const response = await apiClient.post('/api/subscriptions', {
      telegramId: telegramId.toString(),
      shopId,
    });
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка подписки на магазин' };
  }
}

export async function getSubscriptions(telegramId) {
  try {
    const response = await apiClient.get(`/api/subscriptions/${telegramId}`);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка получения подписок' };
  }
}

export async function unsubscribeFromShop(telegramId, shopId) {
  try {
    const response = await apiClient.delete(`/api/subscriptions/${telegramId}/${shopId}`);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка отписки от магазина' };
  }
}

// Payment API functions
export async function generateBitcoinAddress(telegramId) {
  try {
    const response = await apiClient.post('/api/payments/generate-address', {
      telegramId: telegramId.toString(),
      amount: parseFloat(process.env.BITCOIN_SHOP_PRICE) || 25,
    });
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка генерации Bitcoin адреса' };
  }
}

export async function verifyPayment(paymentHash) {
  try {
    const response = await apiClient.post('/api/payments/verify', {
      transactionHash: paymentHash,
    });
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: error.response?.data?.message || 'Ошибка проверки платежа' };
  }
}

export default apiClient;
