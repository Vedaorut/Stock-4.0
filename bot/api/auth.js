import client from './client.js';

/**
 * Authentication API
 */
export const authApi = {
  /**
   * Register new user
   * @param {object} telegramUser - Telegram user object
   * @returns {Promise<object>} User data with JWT token
   */
  async register(telegramUser) {
    return await client.post('/api/auth/register', {
      telegramId: telegramUser.id.toString(),
      username: telegramUser.username,
      firstName: telegramUser.first_name,
      lastName: telegramUser.last_name
    });
  },

  /**
   * Login user via Telegram Web App
   * @param {object} telegramUser - Telegram user object
   * @param {string} initData - Telegram WebApp init data
   * @returns {Promise<object>} User data with JWT token
   */
  async login(telegramUser, initData = null) {
    return await client.post('/api/auth/login', {
      telegramId: telegramUser.id.toString(),
      username: telegramUser.username,
      firstName: telegramUser.first_name,
      lastName: telegramUser.last_name,
      initData
    });
  },

  /**
   * Get current user profile
   * @param {string} token - JWT token
   * @returns {Promise<object>} User profile
   */
  async getProfile(token) {
    return await client.get('/api/auth/profile', token);
  },

  /**
   * Update user profile
   * @param {string} token - JWT token
   * @param {object} updates - Profile updates
   * @returns {Promise<object>} Updated user
   */
  async updateProfile(token, updates) {
    return await client.put('/api/auth/profile', updates, token);
  }
};
