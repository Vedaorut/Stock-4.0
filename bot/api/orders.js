import client from './client.js';

/**
 * Orders API
 */
export const ordersApi = {
  /**
   * Create new order
   * @param {string} token - JWT token
   * @param {object} orderData - Order data
   * @returns {Promise<object>} Created order
   */
  async create(token, orderData) {
    return await client.post('/api/orders', orderData, token);
  },

  /**
   * Get current user's orders
   * @param {string} token - JWT token
   * @param {object} filters - Filter options (status, page, limit)
   * @returns {Promise<object>} Orders list with pagination
   */
  async getMyOrders(token, filters = {}) {
    const params = new URLSearchParams();

    if (filters.status) params.append('status', filters.status);
    if (filters.page) params.append('page', filters.page);
    if (filters.limit) params.append('limit', filters.limit);

    const url = params.toString() ? `/api/orders/my?${params}` : '/api/orders/my';
    return await client.get(url, token);
  },

  /**
   * Get order by ID
   * @param {string} token - JWT token
   * @param {number} orderId - Order ID
   * @returns {Promise<object>} Order details
   */
  async getById(token, orderId) {
    return await client.get(`/api/orders/${orderId}`, token);
  },

  /**
   * Update order status
   * @param {string} token - JWT token
   * @param {number} orderId - Order ID
   * @param {string} status - New status
   * @returns {Promise<object>} Updated order
   */
  async updateStatus(token, orderId, status) {
    return await client.put(`/api/orders/${orderId}/status`, { status }, token);
  }
};
