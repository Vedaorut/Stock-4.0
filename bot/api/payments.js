import client from './client.js';

/**
 * Payments API
 */
export const paymentsApi = {
  /**
   * Verify crypto payment
   * @param {string} token - JWT token
   * @param {number} orderId - Order ID
   * @param {string} txHash - Transaction hash
   * @param {string} currency - Cryptocurrency (BTC, ETH, USDT, TON)
   * @returns {Promise<object>} Verification result
   */
  async verify(token, orderId, txHash, currency) {
    return await client.post('/api/payments/verify', {
      orderId,
      txHash,
      currency
    }, token);
  },

  /**
   * Get payments by order ID
   * @param {string} token - JWT token
   * @param {number} orderId - Order ID
   * @returns {Promise<Array>} Order payments
   */
  async getByOrder(token, orderId) {
    return await client.get(`/api/payments/order/${orderId}`, token);
  },

  /**
   * Check payment status by transaction hash
   * @param {string} token - JWT token
   * @param {string} txHash - Transaction hash
   * @returns {Promise<object>} Payment status
   */
  async checkStatus(token, txHash) {
    return await client.get(`/api/payments/status?txHash=${encodeURIComponent(txHash)}`, token);
  }
};
