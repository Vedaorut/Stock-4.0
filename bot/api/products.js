import client from './client.js';

/**
 * Products API
 */
export const productsApi = {
  /**
   * Create new product
   * @param {string} token - JWT token
   * @param {object} productData - Product data
   * @returns {Promise<object>} Created product
   */
  async create(token, productData) {
    return await client.post('/api/products', productData, token);
  },

  /**
   * List products with filters
   * @param {object} filters - Filter options (shopId, category, inStock, page, limit)
   * @returns {Promise<object>} Products list with pagination
   */
  async list(filters = {}) {
    const params = new URLSearchParams();

    if (filters.shopId) params.append('shopId', filters.shopId);
    if (filters.category) params.append('category', filters.category);
    if (filters.inStock !== undefined) params.append('inStock', filters.inStock);
    if (filters.page) params.append('page', filters.page);
    if (filters.limit) params.append('limit', filters.limit);

    const url = params.toString() ? `/api/products?${params}` : '/api/products';
    return await client.get(url);
  },

  /**
   * Get product by ID
   * @param {number} productId - Product ID
   * @returns {Promise<object>} Product details
   */
  async getById(productId) {
    return await client.get(`/api/products/${productId}`);
  },

  /**
   * Update product
   * @param {string} token - JWT token
   * @param {number} productId - Product ID
   * @param {object} updates - Product updates
   * @returns {Promise<object>} Updated product
   */
  async update(token, productId, updates) {
    return await client.put(`/api/products/${productId}`, updates, token);
  },

  /**
   * Delete product
   * @param {string} token - JWT token
   * @param {number} productId - Product ID
   * @returns {Promise<object>} Delete confirmation
   */
  async delete(token, productId) {
    return await client.delete(`/api/products/${productId}`, token);
  }
};
