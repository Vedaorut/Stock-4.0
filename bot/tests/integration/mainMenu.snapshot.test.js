/**
 * Main Menu Snapshot Test
 *
 * Проверяет что WebApp кнопка:
 * 1. Существует ровно одна
 * 2. Находится в первой строке
 * 3. Нет других URL-кнопок
 */

import { describe, it, expect, beforeEach, afterEach } from '@jest/globals';
import MockAdapter from 'axios-mock-adapter';
import { createTestBot } from '../helpers/testBot.js';
import { callbackUpdate, commandUpdate } from '../helpers/updateFactories.js';
import { extractButtons } from '../helpers/callsCaptor.js';
import { api } from '../../src/utils/api.js';

describe('Main Menu - WebApp Button Position (P0)', () => {
  let testBot;
  let mock;

  beforeEach(() => {
    testBot = createTestBot({ skipAuth: false });
    mock = new MockAdapter(api);

    // Mock auth API
    mock.onPost('/auth/register').reply(200, {
      data: {
        token: 'test-jwt-token',
        user: {
          id: 1,
          telegramId: '123456',
          username: 'testuser',
          selectedRole: null
        }
      }
    });
  });

  afterEach(() => {
    testBot.reset();
    mock.reset();
  });

  it('buyer menu: exactly 1 WebApp button at top, 0 other URL buttons', async () => {
    // Mock buyer has no shop
    mock.onGet('/shops/my').reply(200, { data: [] });

    // Select buyer role
    await testBot.handleUpdate(callbackUpdate('role:buyer'));

    const markup = testBot.getLastMarkup();
    expect(markup).toBeTruthy();

    const buttons = extractButtons(markup);

    // Проверяем количество WebApp кнопок
    const webAppButtons = buttons.filter(b => b.web_app !== undefined);
    expect(webAppButtons.length).toBe(1);

    // Проверяем что WebApp кнопка на первой позиции (minimalist: "Открыть")
    const firstRow = markup.inline_keyboard[0];
    expect(firstRow.length).toBeGreaterThan(0);
    const firstButton = firstRow[0];
    expect(firstButton.web_app).toBeDefined();
    expect(firstButton.text).toContain('Открыть');

    // Проверяем что нет других URL-кнопок (url, login_url)
    const urlButtons = buttons.filter(b => b.url !== undefined || b.login_url !== undefined);
    expect(urlButtons.length).toBe(0);

    // Snapshot markup
    expect(markup).toMatchSnapshot('buyer-menu-no-shop');
  });

  it('seller menu: exactly 1 WebApp button at top, 0 other URL buttons', async () => {
    // Mock seller has shop
    mock.onGet('/shops/my').reply(200, {
      data: [{ id: 1, name: 'Test Shop', sellerId: 1 }]
    });

    // Select seller role
    await testBot.handleUpdate(callbackUpdate('role:seller'));

    const markup = testBot.getLastMarkup();
    expect(markup).toBeTruthy();

    const buttons = extractButtons(markup);

    // Проверяем количество WebApp кнопок
    const webAppButtons = buttons.filter(b => b.web_app !== undefined);
    expect(webAppButtons.length).toBe(1);

    // Проверяем что WebApp кнопка на первой позиции (minimalist)
    const firstRow = markup.inline_keyboard[0];
    expect(firstRow.length).toBeGreaterThan(0);
    const firstButton = firstRow[0];
    expect(firstButton.web_app).toBeDefined();
    expect(firstButton.text).toContain('Открыть');

    // Проверяем что нет других URL-кнопок
    const urlButtons = buttons.filter(b => b.url !== undefined || b.login_url !== undefined);
    expect(urlButtons.length).toBe(0);

    // Snapshot markup
    expect(markup).toMatchSnapshot('seller-menu-with-shop');
  });

  it('seller menu without shop: exactly 1 WebApp button if buyer role opened', async () => {
    // This tests buyerMenuNoShop which also has WebApp button
    
    // Mock seller has no shop
    mock.onGet('/shops/my').reply(200, { data: [] });

    // First select seller to create session
    await testBot.handleUpdate(callbackUpdate('role:seller'));
    testBot.captor.reset();

    // Now toggle to buyer
    await testBot.handleUpdate(callbackUpdate('role:toggle'));

    const markup = testBot.getLastMarkup();
    expect(markup).toBeTruthy();

    const buttons = extractButtons(markup);

    // Проверяем количество WebApp кнопок
    const webAppButtons = buttons.filter(b => b.web_app !== undefined);
    expect(webAppButtons.length).toBe(1);

    // Проверяем что WebApp кнопка на первой позиции
    const firstRow = markup.inline_keyboard[0];
    expect(firstRow.length).toBeGreaterThan(0);
    const firstButton = firstRow[0];
    expect(firstButton.web_app).toBeDefined();

    // Snapshot
    expect(markup).toMatchSnapshot('buyer-menu-after-toggle');
  });

  it('main menu (role selection) has NO WebApp buttons', async () => {
    // User with no saved role sees role selection
    mock.onPost('/auth/register').reply(200, {
      data: {
        token: 'test-jwt-token',
        user: { id: 1, telegramId: '123456', selectedRole: null }
      }
    });

    await testBot.handleUpdate(commandUpdate('start'));

    const markup = testBot.getLastMarkup();
    expect(markup).toBeTruthy();

    const buttons = extractButtons(markup);

    // Проверяем что НЕТ WebApp кнопок (только role:buyer, role:seller)
    const webAppButtons = buttons.filter(b => b.web_app !== undefined);
    expect(webAppButtons.length).toBe(0);

    // Проверяем что есть кнопки выбора роли
    const roleButtons = buttons.filter(
      b => b.callback_data === 'role:buyer' || b.callback_data === 'role:seller'
    );
    expect(roleButtons.length).toBe(2);

    // Snapshot
    expect(markup).toMatchSnapshot('main-menu-role-selection');
  });
});
