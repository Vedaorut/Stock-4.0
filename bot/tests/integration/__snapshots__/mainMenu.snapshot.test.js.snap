// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Menu - WebApp Button Position (P0) buyer menu: exactly 1 WebApp button at top, 0 other URL buttons: buyer-menu-no-shop 1`] = `
{
  "inline_keyboard": [
    [
      {
        "hide": false,
        "text": "📱 Открыть Menu",
        "web_app": {
          "url": "https://your-domain.com",
        },
      },
    ],
    [
      {
        "callback_data": "seller:create_shop",
        "hide": false,
        "text": "➕ Магазин ($25)",
      },
    ],
    [
      {
        "callback_data": "buyer:search",
        "hide": false,
        "text": "🔍 Найти",
      },
    ],
    [
      {
        "callback_data": "buyer:subscriptions",
        "hide": false,
        "text": "📚 Подписки",
      },
    ],
    [
      {
        "callback_data": "buyer:orders",
        "hide": false,
        "text": "🛒 Заказы",
      },
    ],
    [
      {
        "callback_data": "role:toggle",
        "hide": false,
        "text": "🔄 Продавец",
      },
    ],
  ],
}
`;

exports[`Main Menu - WebApp Button Position (P0) main menu (role selection) has NO WebApp buttons: main-menu-role-selection 1`] = `
{
  "inline_keyboard": [
    [
      {
        "callback_data": "role:buyer",
        "hide": false,
        "text": "🛍 Покупатель",
      },
    ],
    [
      {
        "callback_data": "role:seller",
        "hide": false,
        "text": "💼 Продавец",
      },
    ],
  ],
}
`;

exports[`Main Menu - WebApp Button Position (P0) seller menu without shop: exactly 1 WebApp button if buyer role opened: buyer-menu-after-toggle 1`] = `
{
  "inline_keyboard": [
    [
      {
        "hide": false,
        "text": "📱 Открыть Menu",
        "web_app": {
          "url": "https://your-domain.com",
        },
      },
    ],
    [
      {
        "callback_data": "seller:create_shop",
        "hide": false,
        "text": "➕ Магазин ($25)",
      },
    ],
    [
      {
        "callback_data": "buyer:search",
        "hide": false,
        "text": "🔍 Найти",
      },
    ],
    [
      {
        "callback_data": "buyer:subscriptions",
        "hide": false,
        "text": "📚 Подписки",
      },
    ],
    [
      {
        "callback_data": "buyer:orders",
        "hide": false,
        "text": "🛒 Заказы",
      },
    ],
    [
      {
        "callback_data": "role:toggle",
        "hide": false,
        "text": "🔄 Продавец",
      },
    ],
  ],
}
`;

exports[`Main Menu - WebApp Button Position (P0) seller menu: exactly 1 WebApp button at top, 0 other URL buttons: seller-menu-with-shop 1`] = `
{
  "inline_keyboard": [
    [
      {
        "hide": false,
        "text": "📱 Открыть Menu",
        "web_app": {
          "url": "https://your-domain.com",
        },
      },
    ],
    [
      {
        "callback_data": "seller:sales",
        "hide": false,
        "text": "💰 Продажи",
      },
    ],
    [
      {
        "callback_data": "subscription:hub",
        "hide": false,
        "text": "📊 Подписка",
      },
    ],
    [
      {
        "callback_data": "seller:tools",
        "hide": false,
        "text": "🔧 Инструменты",
      },
    ],
    [
      {
        "callback_data": "role:toggle",
        "hide": false,
        "text": "🔄 Покупатель",
      },
    ],
  ],
}
`;
